metadata/filter.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import "strings"
+
+type MetaMap = map[string]string
+
+type metadata struct{}
+
+// Metadata exposes the full implementation
+type Metadata interface {
+	MetaGet
+	MetaCopy
+}
+
+// MetaGet exposes a subset, only Get* methods.
+type MetaGet interface {
+	GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap
+	GetLabelsWithPrefix(src MetaMap, prefix string) MetaMap
+}
+
+// MetaCopy exposes a subset, only Copy* methods.
+type MetaCopy interface {
+	CopyAnnotationsWithPrefix(src, dst MetaMap, prefix string)
+	CopyLabelsWithPrefix(src, dst MetaMap, prefix string)
+}
+
+// NewMetadata is a factory function that return a Metadata interface.
+func NewMetadata() Metadata {
+	return metadata{}
+}
+
+// NewMetaGet is a factory function that return a MetaGet interface.
+func NewMetaGet() MetaGet {
+	return metadata{}
+}
+
+// NewMetaCopy is a factory function that return a MetaCopy interface.
+func NewMetaCopy() MetaCopy {
+	return metadata{}
+}
+
+func (metadata) GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap {
+	return getMetaMapWithPrefix(src, prefix)
+}
+func (metadata) GetLabelsWithPrefix(src MetaMap, prefix string) MetaMap {
+	return getMetaMapWithPrefix(src, prefix)
+}
+
+func (metadata) CopyAnnotationsWithPrefix(src, dst MetaMap, prefix string) {
```

Comment:
Let's keep the copy functions out of this file. This file is for object metadata filtering purposes not to copy/filter generic maps.

Those 4 functions could be replaced with [samber/lo#Assing](https://pkg.go.dev/github.com/samber/lo#Assign) in the `pipeline_run.go` file directly. Something like:
```
lo.Assing(r.Annotations, GetAnnotationsWithPrefix(release)
lo.Assing(r.Labels, GetLabelsWithPrefix(release)
```

Reply:
> Those 4 functions could be replaced with [samber/lo#Assing](https://pkg.go.dev/github.com/samber/lo#Assign)

"assing" :rofl: :laughing: 
I totally get what you mean though, we could "assign" the data another way. 
Or, we could do it right here in this so-called "utility" package.
My feeling is if we are defining tools to get or filter _Labels & Annotations_ by prefix, we may as well assign them too, or at least make it convenient. I don't think we need to add another import. So I'll take a look at doing the merge in the calling code.

Reply:
Take another look.
I've simplified the copy functions in the most recent refactor.
They now receive the `Release` * `PipelineRun` objects directly. 

I don't think we need to export the `Get*` functions at all, but they are still exported, and tested for now.

Reply:
Names are completely different and in my opinion difficult to understand. They are also missing comments, which is a must specially for exported functions.

Reply:
The more recent refactors:

-  adds receivers for `Release` * `PipelineRun` which makes things much better :+1: 
-  removes the `Get*` functions entirely, because in the current form of this patch set, they were not being used at all. 

In my humble opinion, we probably do not need separate `getters` or `setters`, but I'm warm to the idea because that would [conform to how upstream k8s organizes](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/meta.go) everything in `ObjectMeta` with `gets` and `sets` on structure fields. The rational is based on the fact that we wanted to relocate helper functions to copy PAC metadata out of the `tekton` package. Also, an `Add*` method is just a `Copy*` by another name, so it comes down to choosing the name, and I'm indifferent on the names.

> The last two functions would be called with the result of the first ones

That could probably be achieved without much thrash & churn to the current design. So I'll give it a whirl on a local feature branch and see how it goes, and I'll ping you when it's pushed.

Reply:
> That could probably be achieved without much thrash & churn to the current design. So I'll give it a whirl on a local feature branch and see how it goes, and I'll ping you when it's pushed.

Ping @davidmogar the get/add methods were incorporated in the most recent refactor (v). PTAL.